pkg/agent/proxy/proxier.go

```diff
@@ -179,35 +207,38 @@ func (p *proxier) installServices() {
 	for svcPortName, svcPort := range p.serviceMap {
 		svcInfo := svcPort.(*types.ServiceInfo)
 		groupID, _ := p.groupCounter.Get(svcPortName)
-		endpoints, ok := p.endpointsMap[svcPortName]
-		if !ok || len(endpoints) == 0 {
-			continue
-		}
-
-		endpointInstalled, ok := p.endpointInstalledMap[svcPortName]
+		endpointsInstalled, ok := p.endpointInstalledMap[svcPortName]
 		if !ok {
 			p.endpointInstalledMap[svcPortName] = map[string]struct{}{}
-			endpointInstalled = p.endpointInstalledMap[svcPortName]
+			endpointsInstalled = p.endpointInstalledMap[svcPortName]
+		}
+		endpoints := p.endpointsMap[svcPortName]
+		// If both expected Endpoints number and installed Endpoints number are 0, we don't need to take care of this service.
+		if len(endpoints) == 0 && len(endpointsInstalled) == 0 {
+			continue
 		}
 
 		installedSvcPort, ok := p.serviceInstalledMap[svcPortName]
 		var pSvcInfo *types.ServiceInfo
-		needRemoval := false
-		needUpdate := true
+		var needRemoval, needUpdate, needUpdateEndpoints bool
 		if ok {
 			pSvcInfo = installedSvcPort.(*types.ServiceInfo)
 			needRemoval = serviceIdentityChanged(svcInfo, pSvcInfo) || (svcInfo.SessionAffinityType() != pSvcInfo.SessionAffinityType())
 			needUpdate = needRemoval || (svcInfo.StickyMaxAgeSeconds() != pSvcInfo.StickyMaxAgeSeconds())
+			needUpdateEndpoints = pSvcInfo.SessionAffinityType() != svcInfo.SessionAffinityType()
```

Comment:
the code is pretty confusing now: you are basically replacing `needUpdate` with `needUpdateEndpoints`, and you are not using `needUpdate` anywhere in the code below. But you do use `needUpdateEndpoints` as a signal to update the service flows, not just endpoint flows. Therefore, I am not sure why you need a new boolean signal and don't just use the existing `needUpdate`?

Or maybe you just forgot to use `needUpdate` before calling `InstallServiceFlows`?

Reply:
I actually wanted to have finer control here, thus we can avoid updating Endpoints DNAT flows or groups if there are changes that only affect Service flows.
The second thought was that we actually can reach a better control by changing the `needUpdate` to `needUpdateServices`.

Then, we'll continue processing a Service only if it needs either an update on Endpoints datapath or an update on Service datapath.
An Endpoints update includes group/bucket and the Endpoint DNAT flow.
A Service update includes the EndpointSelectionFlow.